x-pack/filebeat/input/s3/collector.go

```diff
@@ -0,0 +1,635 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package s3
+
+import (
+	"bufio"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	awssdk "github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/awserr"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/s3iface"
+	"github.com/aws/aws-sdk-go-v2/service/sqs"
+	"github.com/aws/aws-sdk-go-v2/service/sqs/sqsiface"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+type s3Collector struct {
+	cancellation context.Context
+	logger       *logp.Logger
+
+	config            *config
+	visibilityTimeout int64
+
+	sqs       sqsiface.ClientAPI
+	s3        s3iface.ClientAPI
+	publisher beat.Client
+}
+
+type s3Info struct {
+	name                     string
+	key                      string
+	region                   string
+	arn                      string
+	expandEventListFromField string
+}
+
+type bucket struct {
+	Name string `json:"name"`
+	Arn  string `json:"arn"`
+}
+
+type object struct {
+	Key string `json:"key"`
+}
+
+type s3BucketObject struct {
+	bucket `json:"bucket"`
+	object `json:"object"`
+}
+
+type sqsMessage struct {
+	Records []struct {
+		EventSource string         `json:"eventSource"`
+		AwsRegion   string         `json:"awsRegion"`
+		EventName   string         `json:"eventName"`
+		S3          s3BucketObject `json:"s3"`
+	} `json:"Records"`
+}
+
+type s3Context struct {
+	mux  sync.Mutex
+	refs int
+	err  error // first error witnessed or multi error
+	errC chan error
+}
+
+var (
+	// The maximum number of messages to return. Amazon SQS never returns more messages
+	// than this value (however, fewer messages might be returned).
+	maxNumberOfMessage int64 = 10
+
+	// The duration (in seconds) for which the call waits for a message to arrive
+	// in the queue before returning. If a message is available, the call returns
+	// sooner than WaitTimeSeconds. If no messages are available and the wait time
+	// expires, the call returns successfully with an empty list of messages.
+	waitTimeSecond int64 = 10
+)
+
+func (c *s3Collector) run() {
+	defer c.logger.Infof("s3 input worker for '%v' has stopped.", c.config.QueueURL)
```

Comment:
I've seen multiple instances of adding the queue URL to the log message. Would it make sense to add it as a "contextual" data to the logger like this?
```
c.logger = ctx.Logger.With("queue-url", config.QueueURL)
```
Then you would not have to add this information to all log messages manually.

Reply:
@kvch Good point, this is actually already done. I will remove the queue URLs from the rest of the log messages.